Source/objects.cpp

```diff
@@ -2733,32 +2733,30 @@ void OperateShrineEnchanted(Player &player)
 	if (&player != MyPlayer)
 		return;
 
-	int cnt = 0;
-	uint64_t spell = 1;
-	int maxSpells = gbIsHellfire ? MAX_SPELLS : 37;
-	uint64_t spells = player._pMemSpells;
-	for (int j = 0; j < maxSpells; j++) {
-		if ((spell & spells) != 0)
-			cnt++;
-		spell *= 2;
-	}
-	if (cnt > 1) {
-		spell = 1;
+	const unsigned maxSpells = gbIsHellfire ? MAX_SPELLS : 37;
+
+	const uint64_t knownSpellMask = (0xFFFFFFFFULL >> (64 - maxSpells));
+	if ((player._pMemSpells & knownSpellMask) != 0) {
+		// If the player knows any spells
+		int r;
+		do {
+			// choose a spell at random to decrement
+			r = GenerateRnd(maxSpells) + 1;
+		} while ((player._pMemSpells & GetSpellBitmask(r)) == 0);

```

Comment:
Negative values returned by the RNG I think end up calling lshift with a negative index, resulting in a mask of 0x00000000, so the loop continues.

Reply:
I'm not gonna bother fixing this properly in this PR, going to use the seed directly like in #3314 to pick the spell to decrement. Need to do a bit of cleanup around the way GetSpellBitmask and _pSplLvl is used to make it look nice though.

Reply:
And I did it anyway -_- I actually specified the wrong mask for determining player spells so it was easier to fix it all together.